scylla/src/cluster/state.rs

```
@@ -203,17 +203,22 @@ impl ClusterState {
         &self,
         keyspace: &str,
         table: &str,
-        partition_key: &SerializedValues,
-    ) -> Result<Token, TokenCalculationError> {
-        let partitioner = self
+        partition_key: &dyn SerializeRow,
+    ) -> Result<Token, PartitionKeyError> {
+        let table = self
             .keyspaces
             .get(keyspace)
-            .and_then(|k| k.tables.get(table))
+            .and_then(|k| k.tables.get(table));
+        let values = SerializedValues::from_serializable(
+            &RowSerializationContext::from_specs(table.unwrap().pk_column_specs.as_slice()),
```

This unwrap is not ok. We should probably return an error if we don't have metadata for given `keyspace.table`. Now I wonder if extending `PartitionKeyError` by additional enum is fine, or we should rather introduce new error type. New variant would never be constructed in `PreparedStatement` API - e.g. `PreparedStatement::compute_token()`.

BTW, previously user could provide some garbage keyspace and table names, and we would still compute the token (with the default partitioner - murmur3). I wonder if that was OK.

---

Ooops, forgot about this unwrap, sorry.
I'll let you judge if it is better to introduce new error type or extend PartitionKeyError. I lean slightly towards extending, but it's not a strong opinion.

> BTW, previously user could provide some garbage keyspace and table names, and we would still compute the token (with the default partitioner - murmur3). I wonder if that was OK.

Imo it was not ok. It would default to murmur partitioner, so it would silently break if trying to calculate a token for non-existent CDC table.

---

> Ooops, forgot about this unwrap, sorry. I'll let you judge if it is better to introduce new error type or extend PartitionKeyError. I lean slightly towards extending, but it's not a strong opinion.

Let's extend the `PartitionKeyError`. I don't think there's much sense in introducing the distinction and having two separate types. Both sets of methods are related to computation of token based on PK anyway.

---

Two questions:
- There is `into_execution_error` method on `PartitionKeyError`. I don't really see how to convert this new variant. I don't see how to solve that. Maybe we do need new type after all?
- Should I include the name of the table and keyspace that are missing in the error? Since this is user-facing API this doesn't sound very useful (because user knows what names they passed), and it would require an allocation to store those Strings.

---

> Two questions:
> 
>     * There is `into_execution_error` method on `PartitionKeyError`. I don't really see how to convert this new variant. I don't see how to solve that. Maybe we do need new type after all?

Oh, true. I totally forgot about it. Yeah, I think that this strongly suggests us to introduce a new type (`ClusterStateError`??? - maybe you will come up with some better name). 

>     * Should I include the name of the table and keyspace that are missing in the error? Since this is user-facing API this doesn't sound very useful (because user knows what names they passed), and it would require an allocation to store those Strings.

I'd include them nonetheless. There might be some complex use cases where ks and table names are not passed directly as literals, but are dynamically computed in some manner.